### PR TITLE
refactor(web): browser focus-maintenance states 🧩

### PR DESCRIPTION
**Pull request by jahorton**
This PR serves mostly to spin off focus-maintenance state management from the other bits of functionality in the namespaced `UIManager` class and to provide clearer nomenclature for its fields and methods.

There are two properties from that class that have long given me issue when maintaining related code:
- `isActivating`
- `justActivated`

They are referenced across multiple source files in numerous event handlers, with the design / model merely just "implied" at best.  During development of the es-module (🧩) feature branch so far, I've had to stop, pause, and re-investigate the design on at least 3 separate occasions so far!

If nothing else, it's time to spin the focus-related stuff off as its own module, add _copious_ levels of documentation, and to rename the fields to something that makes their purpose far, far clearer.

- `isActivating` => `maintainingFocus` - generally set when the user is interacting with the OSK.  If the OSK is visible, that means there's an active context... that we don't want to lose.  Thus, it's used to prevent the context from deactivating due to loss-of-page-focus events.
- `justActivated` => `restoringFocus` - generally set immediately before programmatically refocusing a context-providing page element, reactivating it if necessary.  This most frequently occurs when activating a new keyboard, as it's good UX / feedback to display that the keyboard successfully loaded ASAP after the user requests a keyboard change.
- `setActivatingUI()` => `setMaintainingFocus()` - because that's literally all the method ever did, so far as I can tell.  It's basically a setter, just not as a property setter.

The whole point of tracking these properties is to assist KMW's context-management logic with maintaining focus on elements that the user believes they are typing into.  Thus, the object's primary owner will be `app/browser`'s `ContextManager`.

----

Note that this PR is almost entirely, uh, "focused" on the following:
- spinning off these fields and methods into their own module
- renaming them to ease future maintenance
- updating any and all reference points to them, **in place**, with no further related edits...
    - Aside from moving the directly-related API methods to their new site in the module-based engine.

There should be no actual functionality changes here, and any pending hooks (like those documented for OSK events) should still be disconnected.  Those will be a follow-up effort, as I'd like to keep this relatively tight and focused.
- Toward that end, note that the first commit does the code-splitting, with later commits (explicitly named, at that) performing the renames.  This should make the changes _very_ easy to track for review.

@keymanapp-test-bot skip

### CODE REVIEW

File: web/src/app/browser/src/context/focusAssistant.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+/**
+ * The return object documented for
+ * https://help.keyman.com/DEVELOPER/ENGINE/WEB/16.0/reference/core/getUIState.
```
Comment (mcdurdin) [inline web/src/app/browser/src/context/focusAssistant.ts:3]:
```suggestion
 * https://help.keyman.com/developer/engine/web/16.0/reference/core/getUIState.
```

The UPPER/CASE is a bug in IIS, we don't want to propagate that.

File: web/src/app/browser/src/oskConfiguration.ts [hunk 2]
```diff
@@ -19,16 +19,16 @@ export function setupEmbeddedListeners(osk: OSKView) {
...
     if(osk) {
-      keyman.uiManager.setActivatingUI(false);
+      contextManager.focusAssistant.setMaintainingFocus(false);
       osk.startHide(true);
       keyman.domManager.lastActiveElement = null;
     }
   });
 
-  osk.on('onhide', () => {
+  osk.on('onhide', (hiddenByUser) => {
```
Comment (mcdurdin) [inline web/src/app/browser/src/oskConfiguration.ts:28]: This seems like it should be `osk.on('hide'...`?

File: web/src/engine/namespaced-main/dom/domEventHandlers.ts [hunk 3]
```diff
@@ -186,15 +186,15 @@ namespace com.keyman.dom {
       /* If the KeymanWeb UI is active as a user changes controls, all UI-based effects should be restrained to this control in case
       * the user is manually specifying languages on a per-control basis.
       */
-      this.keyman.uiManager.justActivated = false;
+      focusAssistant.restoringFocus = false;
 
-      var isActivating = this.keyman.uiManager.isActivating;
+      var maintainingFocus = focusAssistant.maintainingFocus;
```
Comment (mcdurdin) [inline web/src/engine/namespaced-main/dom/domEventHandlers.ts:191]:
```suggestion
      let maintainingFocus = focusAssistant.maintainingFocus;
```

File: web/src/engine/namespaced-main/dom/domManager.ts [hunk 4]
```diff
@@ -1213,15 +1213,15 @@ namespace com.keyman.dom {
     set activeElement(Pelem: HTMLElement) {
       DOMEventHandlers.states._activeElement = Pelem;
 
-      var isActivating = this.keyman.uiManager.isActivating;
+      var maintainingFocus = focusAssistant.maintainingFocus;
```
Comment (mcdurdin) [inline web/src/engine/namespaced-main/dom/domManager.ts:1216]:
```suggestion
      let maintainingFocus = focusAssistant.maintainingFocus;
```

**Review by mcdurdin — APPROVED** [review]

File: web/src/app/browser/src/oskConfiguration.ts (hunk 2 above)
Comment (jahorton) [reply web/src/app/browser/src/oskConfiguration.ts:28]: Yeah... there's probably a polish pass to be done somewhere in here.

The point of confusion:

https://github.com/keymanapp/keyman/blob/d5f518205fa4180e606d6db74f4c247f468ec8b2/web/src/engine/osk/src/views/oskView.ts#L43-L53

These are _not_ compatible with `.on` - that's a Node-style event-related method.  They instead use the Web-documented `addEventListener`.  (Note:  Node's events also use `addListener`... but not `addEventListener`.  This is generally not ideal, but at least the related type signatures are distinct and so the two types of events never cross within the same method.)

vs

https://github.com/keymanapp/keyman/blob/d5f518205fa4180e606d6db74f4c247f468ec8b2/web/src/engine/osk/src/views/oskView.ts#L60

https://github.com/keymanapp/keyman/blob/d5f518205fa4180e606d6db74f4c247f468ec8b2/web/src/engine/osk/src/views/oskView.ts#L100-L104

https://github.com/keymanapp/keyman/blob/d5f518205fa4180e606d6db74f4c247f468ec8b2/web/src/engine/osk/src/views/oskView.ts#L215

These use the Node-style signature and are thus compatible with `on`.

We obviously still need to maintain the longtime documented KMW events, and I suppose in my haste to have a shinier, Node-based set for an unattached OSK, I'd written a duplicate set using the Node format.  I should probably just eliminate `onhide` and `onshow` for now and just get by with the `hide` and `show` ones.

Though, for what it's worth, the Node-based ones were far more accessible for unit testing up until the point of the previous PR in this chain - #8591.  Before then, I wasn't sure exactly how the implementation for Web's documented events would end up looking once I got to this stage... back when I was modularizing the OSK.  Now that a pattern for the documented events has been established, the duplicated events here will be much easier to resolve.  

Just... probably not in this PR.  Should I mark this as an issue on the feature-branch for cleanup?

Comment (mcdurdin) [reply web/src/app/browser/src/oskConfiguration.ts:28]: > Should I mark this as an issue on the feature-branch for cleanup?

SGTM

**Pull request merged**